test/TokenLock.spec.ts

```diff
@@ -47,4 +54,195 @@ describe('TokenLock', async () => {
+      vest = await tokenLock.vesting(user.address);
+      expect(vest.unlockBegin).to.be.eq(unlockBegin);
+      expect(vest.unlockCliff).to.be.eq(unlockCliff);
+      expect(vest.unlockEnd).to.be.eq(unlockEnd);
+    });
+
+    it('should revert if begin > cliff', async () => {
+      await expect(tokenLock.connect(admin).setupVesting(user.address, 1000, 999, 2000)).to.be.revertedWith(
+        'TokenLock: Unlock cliff must not be before unlock begin'
+      );
+
+      await expect(
+        tokenLock.connect(admin).setupVesting(user.address, unlockBegin, unlockBegin - 1, unlockEnd)
+      ).to.be.revertedWith('TokenLock: Unlock cliff must not be before unlock begin');
+    });
+
+    it('should revert if cliff > end', async () => {
+      await expect(tokenLock.connect(admin).setupVesting(user.address, 1000, 1000, 999)).to.be.revertedWith(
+        'TokenLock: Unlock end must not be before unlock cliff'
+      );
+
+      await expect(
+        tokenLock.connect(admin).setupVesting(user.address, unlockBegin, unlockCliff, unlockCliff - 1)
+      ).to.be.revertedWith('TokenLock: Unlock end must not be before unlock cliff');
+    });
+  });
+
+  describe('#lock', async () => {
+    beforeEach('set vesting schedule for user', async () => {
+      await tokenLock.connect(admin).setupVesting(user.address, unlockBegin, unlockCliff, unlockEnd);
+    });
+
+    it('should revert locks if unlock period has completed', async () => {
+      await setNextBlockTimeStamp(unlockEnd);
+      await expect(tokenLock.connect(admin).lock(user.address, ONE)).to.be.revertedWith(
+        'TokenLock: Unlock period already complete'
+      );
+    });
+
+    it('should revert lock if vesting has not been setup for recipient', async () => {
+      await expect(tokenLock.connect(admin).lock(other.address, ONE)).to.be.revertedWith(
+        'TokenLock: Unlock period already complete'
+      );
+    });
+
+    it('should revert if lock() is called without sufficient balance or allowance', async () => {
+      await expect(tokenLock.connect(user).lock(user.address, ONE)).to.be.revertedWith(
+        'ERC20: transfer amount exceeds balance'
+      );
+      await expect(tokenLock.connect(admin).lock(user.address, ONE)).to.be.revertedWith(
+        'ERC20: transfer amount exceeds allowance'
+      );
+    });
+
+    describe('successful lock', async () => {
+      beforeEach('admin gives tokenLock token allowance', async () => {
+        await token.connect(admin).approve(tokenLock.address, MAX_UINT);
+      });
+
+      it('should emit Locked event', async () => {
+        expect(await tokenLock.connect(admin).lock(user.address, ONE_18))
+          .to.emit(tokenLock, 'Locked')
+          .withArgs(admin.address, user.address, ONE_18);
+      });
+
+      it('should have transferred tokens from caller to contract', async () => {
+        await expect(() => tokenLock.connect(admin).lock(user.address, ONE_18)).to.changeTokenBalances(
```

Comment:
this is cool, didn't know you could do it like that with `changeTokenBalances`

Reply:
yeah it's a neat waffle function =)